compiler: use Tarjan's SCC algorithm to detect loops for defer

niaow opened the pull request:
The compiler needs to know whether a defer is in a loop to determine whether to allocate stack or heap memory. Previously, this performed a DFS of the CFG every time a defer was found. This resulted in time complexity jointly proportional to the number of defers and the number of blocks in the function.

Now, the compiler will instead use Tarjan's strongly connected components algorithm to find cycles in linear time. The search is performed lazily, so this has minimal performance impact on functions without defers.

In order to implement Tarjan's SCC algorithm, additional state needed to be attached to the blocks. I chose to merge all of the per-block state into a single slice to simplify memory management.

niaow committed:
compiler: use Tarjan's SCC algorithm to detect loops for defer

The compiler needs to know whether a defer is in a loop to determine whether to allocate stack or heap memory.
Previously, this performed a DFS of the CFG every time a defer was found.
This resulted in time complexity jointly proportional to the number of defers and the number of blocks in the function.

Now, the compiler will instead use Tarjan's strongly connected components algorithm to find cycles in linear time.
The search is performed lazily, so this has minimal performance impact on functions without defers.

In order to implement Tarjan's SCC algorithm, additional state needed to be attached to the blocks.
I chose to merge all of the per-block state into a single slice to simplify memory management.

niaow commented:
I think I see what is going on here. I seem to have missed it since i ran mostly the non-device tests when developing this. The map seems to have been hiding the assignment with a nil block before.

deadprogram commented on compiler/defer.go:
```diff
@@ -234,41 +234,88 @@ func (b *builder) createInvokeCheckpoint() {
+		}
+		successorNode := &b.blockInfo[successorIndex].tarjan
+
+		if successorNode.index == 0 {
```
I prefer this syntax:

```go
switch {
case successorNode.index == 0:
	// This node has not yet been visited.
	b.strongConnect(successor)
case !successorNode.onStack:
	// This node has been visited, but is in a different SCC.
	// Ignore it.
	continue
}
```

Also note spelling/grammar corrections. :smile_cat:

niaow replied:
Done.

deadprogram commented:
Passes the test corpus, with #5081 applied. At least in those tests there is no apparent performance difference.

@niaow this seems like a good idea overall. Is there a way we can benchmark it?

deadprogram commented:
https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm for the curious.

niaow commented:
I made the requested change, removed `index` from `tarjanState` since we don't actually need it, and added some more comments to explain the used variations.

I don't think there is a good way or need to benchmark this. It only really matters if you have a bunch of defers in a gigantic function. We generally have other stuff that uses a lot more CPU time. I could create some crazy function with a huge number of blocks and defers but that wouldn't really be a good benchmark.

deadprogram commented:
Thanks for the improvement here @niaow now merging.
